### PR TITLE
Upgrade binutils to fix CVE-2021-46174,CVE-2022-44840,CVE-2022-45703,CVE-2022-47007,CVE-2022-47008,CVE-2022-47010,CVE-2022-47011,CVE-2022-47673,CVE-2022-47695,CVE-2022-47696,CVE-2022-48063,CVE-2022-48064,CVE-2022-48065

### PR DESCRIPTION
**Pull request by sindhu-karri**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [ ] The toolchain has been rebuilt successfully (or no changes were made to it)
- [ ] The toolchain/worker package manifests are up-to-date
- [ ] Any updated packages successfully build (or no packages were changed)
- [ ] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [ ] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [ ] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [ ] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [ ] All source files have up-to-date hashes in the `*.signatures.json` files
- [ ] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [ ] Documentation has been updated to match any changes to the build system
- [ ] Ready to merge

---

###### Summary 

What does the PR accomplish, why was it needed?
The PR upgrades binutils from 2.37 to 2.40 to fix 13 CVEs: CVE-2021-46174,CVE-2022-44840,CVE-2022-45703,CVE-2022-47007,CVE-2022-47008,CVE-2022-47010,CVE-2022-47011,CVE-2022-47673,CVE-2022-47695,CVE-2022-47696,CVE-2022-48063,CVE-2022-48064,CVE-2022-48065

###### Change Log  

- Fixed CVE-2021-46174,CVE-2022-44840,CVE-2022-45703,CVE-2022-47007,CVE-2022-47008,CVE-2022-47010,CVE-2022-47011,CVE-2022-47673,CVE-2022-47695,CVE-2022-47696,CVE-2022-48063,CVE-2022-48064,CVE-2022-48065
- Upgraded binutils from 3.27 to 3.40
- Removed old patches that are not required in this version: linker-script-readonly-keyword-support.patch, thin_archive_descriptor.patch, CVE-2021-45078.patch, CVE-2022-38533.patch, CVE-2022-4285.patch

###### Does this affect the toolchain?  

YES

###### Associated issues  

- #xxxx

###### Links to CVEs  
- https://nvd.nist.gov/vuln/detail/CVE-YYYY-XXXX

###### Test Methodology

- Pipeline build id: xxxx
- Buddy build successful: https://dev.azure.com/mariner-org/mariner/_build/results?buildId=415494&view=logs&j=75cba72d-09cb-503a-3927-4ab61ac4a179&t=266a177c-c74b-572c-2e05-3a6c5cd5b38a

### CODE REVIEW

File: SPECS/binutils/binutils.spec [hunk 1]
```diff
@@ -1,7 +1,7 @@
 Summary:        Contains a linker, an assembler, and other tools
 Name:           binutils
-Version:        2.37
-Release:        5%{?dist}
+Version:        2.40
```
Comment (mfrw) [inline SPECS/binutils/binutils.spec:3]: I would be slightly iffy with upgrading `binutils`.
Lets do a full build and ensure things work as we expect them to.
PS: Not the `DELTA_BUILD=y` .. but the full build from scratch & verify with an LKG as well.

File: SPECS/binutils/binutils.spec [hunk 2]
```diff
@@ -130,7 +150,33 @@ sed -i 's/testsuite/ /g' gold/Makefile
...
+%{_libdir}/libsframe.so.0
+%{_libdir}/libsframe.so.0.*
+
+%dir %{_exec_prefix}/lib/debug/%{_libdir}/gprofng
+%{_exec_prefix}/lib/debug/%{_libdir}/gprofng/libgp*
+%{_sysconfdir}/gprofng.rc
+%endif
+
 %changelog
+* Wed Aug 30 2023 Sindhu Karri <lakarri@microsoft.com> - 2.40-1
```
Comment (mfrw) [inline SPECS/binutils/binutils.spec:177]: Also, `binutils` affects the `toolchain` so please update the `manifest` files as well.

Comment (sindhu-karri) [reply SPECS/binutils/binutils.spec:174]: Updated manifests

File: SPECS/binutils/binutils.spec (hunk 1 above)
Comment (sindhu-karri) [reply SPECS/binutils/binutils.spec:3]: full build triggered. https://dev.azure.com/mariner-org/mariner/_build/results?buildId=415761&view=results

File: toolkit/resources/manifests/package/toolchain_x86_64.txt [hunk 3]
```diff
@@ -9,9 +9,9 @@ bash-5.1.8-3.cm2.x86_64.rpm
 bash-debuginfo-5.1.8-3.cm2.x86_64.rpm
 bash-devel-5.1.8-3.cm2.x86_64.rpm
 bash-lang-5.1.8-3.cm2.x86_64.rpm
-binutils-2.37-5.cm2.x86_64.rpm
-binutils-debuginfo-2.37-5.cm2.x86_64.rpm
-binutils-devel-2.37-5.cm2.x86_64.rpm
+binutils-2.40-1.cm2.x86_64.rpm
+binutils-debuginfo-2.40-1.cm2.x86_64.rpm
+binutils-devel-2.40-1.cm2.x86_64.rpm
```
Comment (anphel31) [inline toolkit/resources/manifests/package/toolchain_x86_64.txt:14]: please add the binutils-gprofng subpackage to these toolchain_* manifests

File: SPECS/binutils/linker-script-readonly-keyword-support.patch [hunk 4]
```diff
@@ -1,156 +0,0 @@
-From 6b86da53d5ee2022b9065f445d23356190380746 Mon Sep 17 00:00:00 2001
```
Comment (anphel31) [inline SPECS/binutils/linker-script-readonly-keyword-support.patch:1]: This file should not be removed as it is used with binutils when bootstrapping the toolchain

File: toolkit/resources/manifests/package/toolchain_x86_64.txt (hunk 3 above)
Comment (sindhu-karri) [reply toolkit/resources/manifests/package/toolchain_x86_64.txt:14]: Added

File: SPECS/binutils/linker-script-readonly-keyword-support.patch (hunk 4 above)
Comment (sindhu-karri) [reply SPECS/binutils/linker-script-readonly-keyword-support.patch:1]: This patch is no longer needed in the latest version.

File: SPECS/binutils/binutils.spec (hunk 1 above)
Comment (jslobodzian) [reply SPECS/binutils/binutils.spec:3]: Doublng down on what Falak mentioned.  Can we do this with just patches?

Comment (sindhu-karri) [reply SPECS/binutils/binutils.spec:3]: will check if all patches can be backported.

Comment (sindhu-karri) [reply SPECS/binutils/binutils.spec:3]: raised a new PR with patches for 10 of the 13 CVEs https://github.com/microsoft/CBL-Mariner/pull/6176

**Comment by sindhu-karri** [comment]
Going with backport patches instead of upgrade.
PR at https://github.com/microsoft/CBL-Mariner/pull/6110

**Pull request closed without merging**